rust/hyperlane-core/src/accumulator/historic.rs

```diff
@@ -0,0 +1,155 @@
+    pub fn merge(self, b: MerkleTree) -> MerkleTree {
+        match self {
+            MerkleTree::Zero(_) => self,
+            MerkleTree::Leaf(_) => {
+                if self.hash().eq(&b.hash()) {
+                    b
+                } else {
+                    self
+                }
+            }
+            MerkleTree::Node(a_hash, ref a_left, ref a_right) => match b {
+                MerkleTree::Leaf(_) => self,
+                MerkleTree::Zero(_) => self,
+                MerkleTree::Node(_, b_left, b_right) => {
+                    let merged_left = (**a_left).clone().merge((*b_left).clone());
+                    let merged_right = (**a_right).clone().merge((*b_right).clone());
+                    let merged_hash = hash_concat(merged_left.hash(), merged_right.hash());
+                    assert_eq!(merged_hash, a_hash);
+                    MerkleTree::Node(a_hash, Box::new(merged_left), Box::new(merged_right))
+                }
+            },
+        }
+    }
+}
+
+impl Proof {
+    /// Return the proof of this index when it was the latest node in the tree
+    pub fn as_latest(&self) -> Proof {
+        // Replace the right nodes with zero hashes
+        let mut modified_path = [H256::zero(); TREE_DEPTH];
+        for i in 0..TREE_DEPTH {
+            let size = self.index >> i;
+            if (size & 1) == 1 {
+                modified_path[i] = self.path[i];
+            } else {
+                modified_path[i] = ZERO_HASHES[i];
+            }
+        }
+
+        Proof {
+            leaf: self.leaf,
+            index: self.index,
+            path: modified_path,
+        }
+    }
+
+    /// Creates a partial merkle tree out of the proof
+    pub fn partial_tree(&self) -> MerkleTree {
+        let mut tree = MerkleTree::Leaf(self.leaf);
+
+        for i in 0..TREE_DEPTH {
+            let index = self.index >> i;
+            if (index & 1) == 1 {
+                let left = MerkleTree::Leaf(self.path[i]);
+                let hash = hash_concat(left.hash(), tree.hash());
+                tree = MerkleTree::Node(hash, Box::new(left), Box::new(tree));
+            } else {
+                let right = MerkleTree::Leaf(self.path[i]);
+                let hash = hash_concat(tree.hash(), right.hash());
+                tree = MerkleTree::Node(hash, Box::new(tree), Box::new(right));
+            }
+        }
+        tree
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::accumulator::{
+        merkle::{verify_merkle_proof, MerkleTree},
+        TREE_DEPTH,
+    };
+
+    use super::*;
+    fn generate_proof(tree: &MerkleTree, index: usize) -> Proof {
+        // Generate a proof of index i from the full merkle tree
+        let (leaf, hashes) = tree.generate_proof(index, TREE_DEPTH);
+        let mut path = [H256::zero(); TREE_DEPTH];
+        path.copy_from_slice(&hashes[..TREE_DEPTH]);
+        Proof { index, path, leaf }
+    }
+
+    #[test]
+    fn as_latest() {
```

Comment:
Great test, very thorough

May want to rename this, it wasn't clear to me at first that this also covers merge / partial_tree

Reply:
Definitely should rename